### PR TITLE
Likes Fix (button mashing and initial negative likes)

### LINKED ISSUE #3524: Continuous like button press shows unexpected behavior
### Please describe the problem (or idea)
When like button is pressed continuously it shows unexpected behaviour. Like may go to negative values also. On refreshing page problem is solved.

> What happened just before the problem occurred? Or what problem could this idea solve?

like

> What did you expect to see that you didn't?

### Please show us where to look

https://publiclab.org/notes/warren/10-31-2016/create-a-welcoming-first-timers-only-issue-to-invite-new-software-contributors

### What's your PublicLab.org username?
igniteeng000
> This can help us diagnose the issue: 

### Browser, version, and operating system
Windows 10, chrome
> Many bugs are related to these -- please help us track it down and reproduce what you're seeing!

****

## Thank you!

Your help makes Public Lab better! We *deeply* appreciate your helping refine and improve this site. 

To learn how to write really great issues, which increases the chances they'll be resolved, see:

https://publiclab.org/wiki/developers#Contributing+for+non-coders

### PR DESCRIPTION
**Pull request by sashadev-sky**
Fixes #3524, #4589

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [x] PR is descriptively titled 📑
* [x] screenshots/GIFs are attached 📎 in case of UI updation
* [x] ask `@publiclab/reviewers` for help, in a comment below

================

**Please find notes and GIFs here**: 

[https://github.com/sashadev-sky/PL-notes/blob/master/likes.md](https://github.com/sashadev-sky/PL-notes/blob/master/likes.md)

===============

### CODE REVIEW

**Comment by sashadev-sky** [comment]
@jywarren From working with this code I have felt like the notification popups UI isn't compatible with the  'like' and 'unlike' functionalities. I was thinking maybe to just remove them all together as they might be replaced when a notifications API is fully integrated with the site. Please see my comments and GIFs in the link in my PR description and let me know your thoughts! Thank you

**Comment by avsingh999** [comment]
@sashadev-sky  can you please add Gif 
thanks : )

**Comment by sashadev-sky** [comment]
@avsingh999 the Gif is linked in the PR request description. It wouldn't let me drag it here because the file was too large

**Comment by avsingh999** [comment]
> @avsingh999 the Gif is linked in the PR request description. It wouldn't let me drag it here because the file was too large

Sorry I didn't see :smile:

**Comment by namangupta01** [comment]
Hi @sashadev-sky, the right approach for doing this according to me is update the likes when we get the response with the number of likes. And disable the button when the request is taking place of the previous like. What do you think?

**Comment by sashadev-sky** [comment]
@namangupta01 I think disabling a button does not prevent event listeners from firing. So that is why I had to make sure to turn them off in between promises

**Comment by namangupta01** [comment]
By disabling the button, I mean to stop progressing the event when request is going on using `preventDefault` or by using `if` `else`.

File: Gemfile [hunk 1]
```diff
@@ -69,20 +69,20 @@ gem 'mailman', require: false
 
 # To convert html to markdown
 gem 'reverse_markdown'
-
+gem 'mysql2', '>= 0.4.4'
```
Comment (namangupta01) [inline Gemfile]: We use quite for Development so please change this back to production env

File: Gemfile [hunk 2]
```diff
@@ -69,20 +69,20 @@ gem 'mailman', require: false
...
 # run with `bundle install --without production` or `bundle install --without mysql` to exclude this
-group :mysql, :production do
-  gem 'mysql2', '>= 0.4.4'
-  # mysql 0.4.3+ causes a version mismatch, apparently, and demands 'activerecord-mysql2-adapter'
-end
+# group :mysql, :production do
```
Comment (namangupta01) [inline Gemfile]: Undo these changes also.

File: Gemfile [hunk 3]
```diff
@@ -69,20 +69,20 @@ gem 'mailman', require: false
...
+#   # mysql 0.4.3+ causes a version mismatch, apparently, and demands 'activerecord-mysql2-adapter'
+# end
 
 group :sqlite, :development do
   gem 'sqlite3', '~> 1.3.6'
 end
 
 gem 'coffee-rails', '~> 4.2.2'
 gem 'execjs' # See https://github.com/sstephenson/execjs#readme for more supported runtimes
-gem 'sass-rails', '~> 5.0', '>= 5.0.7'
```
Comment (namangupta01) [inline Gemfile]: Please change this  version back to original one.

File: app/views/like/_like.html.erb [hunk 4]
```diff
@@ -94,18 +94,22 @@
...
+      // triggers liking an unliking
       <% if current_user %>
         $('#like-button-<%= node.id %>').on('click',<% if node.liked_by(current_user.uid) %>clickliked<% else %>clicknotliked<% end %>)
-      
-        if(window.location.href.includes("#like")) {
-          $('#like-button-<%= node.id %>').click();
-        }
+       
+       // note useful for comments on notes but maybe useful for something else. 
+        // if(window.location.href.includes("#like")) {
```
Comment (namangupta01) [inline app/views/like/_like.html.erb]: It would be better to remove unused code.

File: spec/javascripts/plots2_spec.js [hunk 5]
```diff
@@ -1,92 +1,10 @@
-//= require comment_expand
-//= require like
```
Comment (namangupta01) [inline spec/javascripts/plots2_spec.js]: Is it necessary to like file?

File: app/assets/javascripts/like.js [hunk 6]
```diff
@@ -1,58 +1,39 @@
...
+      updateLikeCount(parseInt(resp), node_id);
+      renderLikeStar(parseInt(resp), node_id);
+    })
+    .then(function(resp) {
+      let method1 = method === "/delete" ? clicknotliked : clickliked
+      $('#like-button-' + node_id).on('click', method1);
+    });
+}
 
+function updateLikeCount(value, node_id) {
```
Comment (namangupta01) [inline app/assets/javascripts/like.js]: please add a test for this

File: app/assets/javascripts/like.js [hunk 7]
```diff
@@ -1,58 +1,39 @@
...
+  var count = $('#like-count-' + node_id).html();
+  count = parseInt(count) + value;
+  $('#like-count-' + node_id).html(count);
 }
+
+// where fa fa-star-o is a clear star (indicating you are not currently liking)
+function renderLikeStar(value, node_id) {
```
Comment (namangupta01) [inline app/assets/javascripts/like.js]: And also please add tests for newly added functions

**Comment by sashadev-sky** [comment]
@namangupta01 this pull request is marked as “in progress” I will make those updates when I’m ready for merge review 👍🏻

**Comment by namangupta01** [comment]
No problem, let me know if you need some help! 👍

File: app/views/like/_like.html.erb (hunk 4 above)
Comment (sashadev-sky) [reply app/views/like/_like.html.erb]: I don’t think code should be deleted before you know the reason it was put there in the first place. As noted in its comment I am still investigating that and this PR is not labeled for review. If you have input on the function of it that would be useful. Thank you

**Comment by sashadev-sky** [comment]
@namangupta01 I don't quite understand the change you are suggesting. I don't think preventdefault should have any effect on the button in this scenario as it is not a form submission

**Comment by jywarren** [comment]
Hi @sashadev-sky - by `front end tests` do you mean front-end javascript tests? We do some JavaScript tests here: https://github.com/publiclab/plots2/tree/master/spec/javascripts but they're not great because they are only run against static fixtures, which need to be kept up to date. There's a little about this here:

https://github.com/publiclab/plots2/blob/master/doc/TESTING.md#client-side-tests

I'm also interested in exploring system tests, in https://github.com/publiclab/plots2/issues/3683, and i just opened a PR here: https://github.com/publiclab/plots2/pull/4888 (though not sure if it will run). 

That will actually run a full-stack version of the site in Chrome, and allows click interactions, so it really tests the whole system. I think this could be a good way forward for testing likes and we may want to move away from the teaspoon/mocha tests since they can't test on dynamic data, and require the static fixtures to be maintained. What do you think?

File: app/views/like/_like.html.erb (hunk 4 above)
Comment (jywarren) [reply app/views/like/_like.html.erb]: You could look through the blame view for this file to see when it was added and when removed, and see if you can track it that way!

File: app/views/notes/_notes.html.erb [hunk 8]
```diff
@@ -34,7 +34,7 @@
 	    <%= distance_of_time_in_words(Time.at(node.latest.timestamp), Time.current, { include_seconds: false, scope: 'datetime.time_ago_in_words' }) %>
           <% end %>
           | <i class="fa fa-eye"></i> <%= number_with_delimiter(node.totalviews) %> <span class="hidden-xs hidden-sm"><%= t('notes._notes.views') %></span>
-          | <i style="<% if node.likes > 0 %>color:#db4;<% else %>color:#888;<% end %>" class="fa fa-star-o"></i> <%= node.likes %>
+          | <i style="<% if node.likers.length > 0 %>color:#db4;<% else %>color:#888;<% end %>" class="fa fa-star-o"></i> <%= node.likers.length %>
```
Comment (jywarren) [inline app/views/notes/_notes.html.erb]: we might do `likers.count` so that it runs a COUNT sql instead of calling the whole record and then counting it.

File: package.json [hunk 9]
```diff
@@ -22,7 +22,7 @@
   },
   "dependencies": {
     "blob.js": "^1.0.1",
-    "blueimp-file-upload": "blueimp/jQuery-File-Upload#^9.22.1",
+    "blueimp-file-upload": "blueimp/jQuery-File-Upload#~8.1.0",
```
Comment (jywarren) [inline package.json]: oops, downgrading this, but I assume you'll remove this before final merge, thanks!

File: spec/javascripts/like_spec.js [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+       // jQuery.getJSON("/likes/node/1/delete").done(function(data){
+       //   response = data;
+       // })
+
+      $("#like-button-1").click();
+      expect($('#like-count-1').html()).to.eql('0');
+      
+  });
+
+  it("it toggles the like star render", function () {
```
Comment (jywarren) [inline spec/javascripts/like_spec.js]: Wow, these are great! Yeah, you can see the issue with static fixtures is not awesome. But great work!

File: app/views/notes/_notes.html.erb (hunk 8 above)
Comment (sashadev-sky) [reply app/views/notes/_notes.html.erb]: @jywarren I made this update and committed just now, but then I remembered I had considered it at some point earlier and left it as `length`. So I double checked the rails logs and I think because plots2 implements cached likes for nodes, it doesn't make a difference if we use `length` or `count` for this specific view (its always the exact same number of queries with similar times on average). I was thinking for consistency with `_likes.html.erb` to leave it as `length`? 

I reverted it back to `length` for now but let me know your thoughts!

File: app/views/like/_like.html.erb (hunk 4 above)
Comment (sashadev-sky) [reply app/views/like/_like.html.erb]: It actually doesn't do anything helpful I looked into it. But instead of removing it I will comment it out. I think I will open a new issue after this for implementing likes on comments (currently not working) and fix it up there. Is that okay with everyone? So the uncommented code will just be there for a bit.

**Comment by sashadev-sky** [comment]
@jywarren thank you for linking to me to the capybara testing issue! As I mentioned there definitely excited about getting that to work but at the risk of getting into a google search black hole looking for work arounds I think maybe for this issue we can just stick with the current testing scheme and open up a new issue once rack is implemented to fix all of the JS frontend tests (there aren't many). What are your thoughts?

2) Also please note my comment on the `.count` and `.length` issue and let me know your thoughts!

3) Last thing: I think this got buried in the comments here but when I initially added this PR I wanted you to advise one thing: 
@jywarren From working with this code I have felt like the notification popups UI isn't compatible with the 'like' and 'unlike' functionalities. I was thinking maybe to just remove them all together as they might be replaced when a notifications API is fully integrated with the site. Please see my comments and GIFs in the link in my PR description and let me know your thoughts! Thank you

Linking GIF here for your convenience:
https://github.com/sashadev-sky/PL-notes/blob/master/likes.md

Thank you!! Looking to get this issue closed finally

File: spec/javascripts/like_spec.js (hunk 10 above)
Comment (sashadev-sky) [reply spec/javascripts/like_spec.js]: Thank you!

**Comment by jywarren** [comment]
Spinner idea: `<i class="fa fa-spinner fa-spin"></i>`

**Comment by jywarren** [comment]
Folks have asked about what debounce means and how this strategy works. @sashadev-sky here's a link to a great explanation: https://css-tricks.com/debouncing-throttling-explained-examples/ -- it can really help throttle client/server interactions and make things smoother. I recently implemented it in this PR with @aashna27 's help: https://github.com/publiclab/plots2/pull/4904/

Example of debounce: https://github.com/publiclab/plots2/pull/4904/files#diff-128f89a0762ba84f1ad9619753645287R15

**Comment by jywarren** [comment]
https://css-tricks.com/the-difference-between-throttling-and-debouncing/

**Comment by jywarren** [comment]
```js
$('.myButton').click(debounce(function onClick(){
   // ... do things
}, 350);
```

**Comment by jywarren** [comment]
https://css-tricks.com/debouncing-throttling-explained-examples/

**Comment by jywarren** [comment]
`immediate` -- https://github.com/component/debounce#debouncefn-wait--immediate--false-

**Comment by sashadev-sky** [comment]
@jywarren likes are debounced! Please review. Not much update to show in terms of UI so the gif in the description is basically the same still as what I saw when running it on the unstable branch. I also saw from the waterfall in Chrome dev tools that the debounce is taking pauses for the requests and not every click will lead to a request when mashing the button.

**Comment by jywarren** [comment]
Awesome and well done here!!!
